### PR TITLE
Feature/archiving vedtak

### PR DESCRIPTION
**Pull request by MengZhuNAV**
archive barnetrydg vedtak use familie-integrasjoner

### CODE REVIEW

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonException.kt [hunk 1]
```diff
@@ -12,7 +12,7 @@ class IntegrasjonException : RuntimeException {
     private var responseBody: ByteArray? = null
 
     constructor(msg: String?) : super(msg)
-    constructor(msg: String?, e: Exception?, uri: URI?, ident: String?) : super(msg, e) {
+    constructor(msg: String?, e: Throwable?, uri: URI?, ident: String?) : super(msg, e) {
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonException.kt]: What is the reason for this change?

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 2]
```diff
@@ -63,7 +70,11 @@ class IntegrasjonTjeneste (
 
     @Retryable(value = [IntegrasjonException::class], maxAttempts = 3, backoff = Backoff(delay = 5000))
     fun journalFørVedtaksbrev(pdf: ByteArray, fnr: String, callback: (journalpostID: String) -> Unit) {
-        callback("journalpostID: TODO")
+        val journalpostId= lagerJournalpostForVedtaksbrev(fnr, pdf)
+        if(journalpostId== null){
+            throw IntegrasjonException("Kall mot integrasjon feilet ved lager journalpost. fnr: ${fnr}")
+        }
+        callback(journalpostId)
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Can we just return the value and loose the complexity of a return function?

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 3]
```diff
@@ -63,7 +70,11 @@ class IntegrasjonTjeneste (
 
     @Retryable(value = [IntegrasjonException::class], maxAttempts = 3, backoff = Backoff(delay = 5000))
     fun journalFørVedtaksbrev(pdf: ByteArray, fnr: String, callback: (journalpostID: String) -> Unit) {
-        callback("journalpostID: TODO")
+        val journalpostId= lagerJournalpostForVedtaksbrev(fnr, pdf)
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: As Bent noted I would also like to see consistent formatting. Especially space before and after `=`.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 4]
```diff
@@ -84,9 +95,38 @@ class IntegrasjonTjeneste (
         }
     }
 
+    fun lagerJournalpostForVedtaksbrev(fnr: String, pdfByteArray: ByteArray): String?{
+        val uri = URI.create("$integrasjonerServiceUri/arkiv/v1")
+        logger.info("Sender vedtak pdf til DokArkiv: ${uri}");
+
+        return runCatching<ArkiverDokumentResponse> {
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Proposal:
Use the function onSuccess to return the value. lagerJournalpostForVedtaksbrev should return String and not String?.

Also I think it is easier to read :)

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 5]
```diff
@@ -84,9 +95,38 @@ class IntegrasjonTjeneste (
...
+            val arkiverDokumentResponse= sendJournalFørRequest(uri, arkiverDokumentRequest)
+            arkiverDokumentResponse
+        }.onFailure {
+            throw IntegrasjonException("Kall mot integrasjon feilet ved lager journalpost.", it, uri, fnr)
+        }.getOrNull()?.journalpostId
+    }
+
+    private fun sendJournalFørRequest(journalFørEndpoint : URI, arkiverDokumentRequest: ArkiverDokumentRequest): ArkiverDokumentResponse {
+        val headers = HttpHeaders()
+        headers.contentType = MediaType.APPLICATION_JSON
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Add utf-8 support in case we need it in the future. There are some examples in the code with utf-8 content type.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 6]
```diff
@@ -84,9 +95,38 @@ class IntegrasjonTjeneste (
...
+        headers.add(NavHttpHeaders.NAV_CALL_ID.asString(), MDC.get(MDCConstants.MDC_CALL_ID))
+        val response= restTemplate.exchange(journalFørEndpoint, HttpMethod.POST, HttpEntity<Any>(arkiverDokumentRequest, headers), Ressurs::class.java)
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: If you use restOperations and remove Ressurs::class.java you dont need the mapper.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 7]
```diff
@@ -84,9 +95,38 @@ class IntegrasjonTjeneste (
...
+        return mapper.convertValue(response.body!!.data, ArkiverDokumentResponse::class.java)
+    }
+
     companion object {
         private val logger = LoggerFactory.getLogger(IntegrasjonTjeneste::class.java)
         private val secureLogger = LoggerFactory.getLogger("secureLogger")
-        private const val OAUTH2_CLIENT_CONFIG_KEY = "familie-integrasjoner-clientcredentials"
+
+        val VEDTAK_FILTYPE= FilType.PDFA
+        const val VEDTAK_FILNAVN= "ba_vb.pdf"
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Have you checked with Fredrik or Birgitte that this is the filename they want?

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt [hunk 8]
```diff
@@ -84,9 +95,38 @@ class IntegrasjonTjeneste (
...
+        const val VEDTAK_DOKUMENT_TYPE= "BARNETRYGD_VEDTAK"
+
+        //for testability make it visible
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: What do you mean?

File: src/test/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjenesteTest.kt [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    val mockServiceUri= "mock.integrasjoner.uri"
+    val mockFnr= "12345678910"
+    val mockPdf= "pdf data".toByteArray()
+
+    @BeforeEach
+    fun setUp()= MockKAnnotations.init(this)
+
+    @Test
+    fun `test å lager journalpost for vedtaksbrev`(){
```
Comment (henninghaakonsen) [inline src/test/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjenesteTest.kt]: For consistency I would recommend using the setup for the other integrationtests in this app. Have a look at ØkonomiIntegrasjonTest for example.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonException.kt (hunk 1 above)
Comment (MengZhuNAV) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonException.kt]: when you use runCatching, the onFailure only provides a Throwable. If we want to throw an IntegrasjonException in onFailure, it is necessary to change the interface.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 2 above)
Comment (MengZhuNAV) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: This method is defined by @tnarland 
I think the complexity has to be addressed somewhere, and lagerJournalpostForVedtaksbrev is more like a "pure" method to do the archiving. This method is used to hook up with Tor's async task mechanism, and I think it is fine

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 4 above)
Comment (MengZhuNAV) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Ok, but the code has to be like this:
`    fun lagerJournalpostForVedtaksbrev(fnr: String, pdfByteArray: ByteArray): String?{
        val uri = URI.create("$integrasjonerServiceUri/arkiv/v1")
        logger.info("Sender vedtak pdf til DokArkiv: ${uri}");

        runCatching<ArkiverDokumentResponse> {
            val dokumenter= listOf(Dokument(pdfByteArray, VEDTAK_FILTYPE, VEDTAK_FILNAVN, null, VEDTAK_DOKUMENT_TYPE))
            val arkiverDokumentRequest= ArkiverDokumentRequest(fnr, true, dokumenter)
            val arkiverDokumentResponse= sendJournalFørRequest(uri, arkiverDokumentRequest)
            arkiverDokumentResponse
        }.onFailure {
            throw IntegrasjonException("Kall mot integrasjon feilet ved lager journalpost.", it, uri, fnr)
        }.onSuccess {
            return it.journalpostId
        }
        
        return null
    }`

the "return null" is necessary because all paths must return a value, which also prevents the function returns String

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 5 above)
Comment (MengZhuNAV) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: done!

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 7 above)
Comment (MengZhuNAV) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: No, I will do it.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 8 above)
Comment (MengZhuNAV) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: the static field was private, but I need it in the tests so I make it visible to test classes

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonException.kt (hunk 1 above)
Comment (henninghaakonsen) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonException.kt]: Ok, maybe also rename the variable to throwable?

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 2 above)
Comment (henninghaakonsen) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: I think this adds complexity for the team and we dont use it in any other cases. I know that this is normal practice in eg. javascript.

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 8 above)
Comment (henninghaakonsen) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Why do the test need the oauth config key?

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 4 above)
Comment (henninghaakonsen) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: Suggestion:
`
    fun lagerJournalpostForVedtaksbrev(fnr: String, pdfByteArray: ByteArray): String {
        val uri = URI.create("$integrasjonerServiceUri/arkiv/v1")
        logger.info("Sender vedtak pdf til DokArkiv: $uri");

        val dokumenter= listOf(Dokument(pdfByteArray, VEDTAK_FILTYPE, VEDTAK_FILNAVN, null, VEDTAK_DOKUMENT_TYPE))
        val arkiverDokumentRequest = ArkiverDokumentRequest(fnr, true, dokumenter)

        return Result.runCatching { sendJournalFørRequest(uri, arkiverDokumentRequest) }
                .fold(
                    onSuccess = {
                        it.journalpostId
                    },
                    onFailure = {
                        throw IntegrasjonException("Kall mot integrasjon feilet ved lager journalpost.", it, uri, fnr)
                    }
                )
    }
`

File: src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt (hunk 2 above)
Comment (tnarland) [reply src/main/kotlin/no/nav/familie/ba/sak/integrasjoner/IntegrasjonTjeneste.kt]: I guess we could simply return the value. Don't know what I was thinking 😆

**Comment by henninghaakonsen** [comment]
I have updated some of the code so that Meng wont have to work on holidays 👍 I suggest @tnarland takes over now and makes it work with familie-integrasjoner.
